File: content/develop/use-cases/leaderboard/go/leaderboard.go
package leaderboard

import (
	"context"
	"fmt"

	"github.com/redis/go-redis/v9"
)

// Config holds the RedisLeaderboard configuration.
type Config struct {
	Client     *redis.Client
	Key        string
	MaxEntries int
}

// Entry is a hydrated leaderboard entry with rank, score, and metadata.
type Entry struct {
	Rank           int               `json:"rank"`
	UserID         string            `json:"user_id"`
	Score          float64           `json:"score"`
	Metadata       map[string]string `json:"metadata"`
	TrimmedUserIDs []string          `json:"trimmed_user_ids,omitempty"`
}

// RedisLeaderboard stores leaderboard data in a sorted set and user metadata in hashes.
type RedisLeaderboard struct {
	client     *redis.Client
	key        string
	maxEntries int
}

// NewRedisLeaderboard creates a leaderboard with sane defaults.
func NewRedisLeaderboard(cfg Config) *RedisLeaderboard {
	key := cfg.Key
	if key == "" {
		key = "leaderboard:demo"
	}

	maxEntries := cfg.MaxEntries
	if maxEntries == 0 {
		maxEntries = 100
	}

	return &RedisLeaderboard{
		client:     cfg.Client,
		key:        key,
		maxEntries: maxEntries,
	}
}

func normalizePositiveInt(value int, fieldName string) (int, error) {
	if value < 1 {
		return 0, fmt.Errorf("%s must be at least 1", fieldName)
	}
	return value, nil
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func (lb *RedisLeaderboard) metadataKey(userID string) string {
	return fmt.Sprintf("%s:user:%s", lb.key, userID)
}

func (lb *RedisLeaderboard) coerceMetadata(metadata map[string]string) map[string]string {
	if metadata == nil {
		return map[string]string{}
	}

	payload := make(map[string]string, len(metadata))
	for field, value := range metadata {
		payload[field] = value
	}
	return payload
}

func (lb *RedisLeaderboard) metadataPayload(metadata map[string]string) map[string]any {
	payload := make(map[string]any, len(metadata))
	for field, value := range metadata {
		payload[field] = value
	}
	return payload
}

func (lb *RedisLeaderboard) deleteMetadataForUsers(ctx context.Context, userIDs []string) error {
	if len(userIDs) == 0 {
		return nil
	}

	keys := make([]string, 0, len(userIDs))
	for _, userID := range userIDs {
		keys = append(keys, lb.metadataKey(userID))
	}

	return lb.client.Del(ctx, keys...).Err()
}

func (lb *RedisLeaderboard) hydrateEntries(
	ctx context.Context,
	entries []redis.Z,
	startRank int,
) ([]Entry, error) {
	if len(entries) == 0 {
		return []Entry{}, nil
	}

	pipe := lb.client.Pipeline()
	metadataCommands := make([]*redis.MapStringStringCmd, 0, len(entries))
	for _, entry := range entries {
		userID, _ := entry.Member.(string)
		metadataCommands = append(metadataCommands, pipe.HGetAll(ctx, lb.metadataKey(userID)))
	}

	if _, err := pipe.Exec(ctx); err != nil {
		return nil, fmt.Errorf("failed to load leaderboard metadata: %w", err)
	}

	hydratedEntries := make([]Entry, 0, len(entries))
	for index, entry := range entries {
		userID, _ := entry.Member.(string)
		metadata, err := metadataCommands[index].Result()
		if err != nil {
			return nil, fmt.Errorf("failed to read metadata for %s: %w", userID, err)
		}

		hydratedEntries = append(hydratedEntries, Entry{
			Rank:     startRank + index,
			UserID:   userID,
			Score:    entry.Score,
			Metadata: metadata,
		})
	}

	return hydratedEntries, nil
}

func (lb *RedisLeaderboard) trimToMaxEntries(ctx context.Context) ([]string, error) {
	size, err := lb.client.ZCard(ctx, lb.key).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get leaderboard size: %w", err)
	}

	overflow := int(size) - lb.maxEntries
	if overflow <= 0 {
		return []string{}, nil
	}

	trimmedUserIDs, err := lb.client.ZRange(ctx, lb.key, 0, int64(overflow-1)).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to fetch trimmed user ids: %w", err)
	}
	if len(trimmedUserIDs) == 0 {
		return []string{}, nil
	}

	if err := lb.client.ZRemRangeByRank(ctx, lb.key, 0, int64(overflow-1)).Err(); err != nil {
		return nil, fmt.Errorf("failed to trim leaderboard: %w", err)
	}

	if err := lb.deleteMetadataForUsers(ctx, trimmedUserIDs); err != nil {
		return nil, err
	}

	return trimmedUserIDs, nil
}

// UpsertUser creates or updates a leaderboard entry and associated metadata.
func (lb *RedisLeaderboard) UpsertUser(
	ctx context.Context,
	userID string,
	score float64,
	metadata map[string]string,
) (*Entry, error) {
	payload := lb.coerceMetadata(metadata)

	pipe := lb.client.TxPipeline()
	pipe.ZAdd(ctx, lb.key, redis.Z{
		Score:  score,
		Member: userID,
	})
	if len(payload) > 0 {
		pipe.HSet(ctx, lb.metadataKey(userID), lb.metadataPayload(payload))
	}
	if _, err := pipe.Exec(ctx); err != nil {
		return nil, fmt.Errorf("failed to upsert leaderboard user: %w", err)
	}

	trimmedUserIDs, err := lb.trimToMaxEntries(ctx)
	if err != nil {
		return nil, err
	}

	entry, err := lb.GetUserEntry(ctx, userID)
	if err != nil {
		return nil, err
	}
	if entry == nil {
		return &Entry{
			UserID:         userID,
			Score:          score,
			Metadata:       payload,
			TrimmedUserIDs: trimmedUserIDs,
		}, nil
	}

	entry.TrimmedUserIDs = trimmedUserIDs
	return entry, nil
}

// IncrementScore increments a user's score and optionally updates metadata.
func (lb *RedisLeaderboard) IncrementScore(
	ctx context.Context,
	userID string,
	amount float64,
	metadata map[string]string,
) (*Entry, error) {
	payload := lb.coerceMetadata(metadata)

	pipe := lb.client.TxPipeline()
	incrCmd := pipe.ZIncrBy(ctx, lb.key, amount, userID)
	if len(payload) > 0 {
		pipe.HSet(ctx, lb.metadataKey(userID), lb.metadataPayload(payload))
	}
	if _, err := pipe.Exec(ctx); err != nil {
		return nil, fmt.Errorf("failed to increment leaderboard score: %w", err)
	}

	newScore, err := incrCmd.Result()
	if err != nil {
		return nil, fmt.Errorf("failed to read incremented score: %w", err)
	}

	trimmedUserIDs, err := lb.trimToMaxEntries(ctx)
	if err != nil {
		return nil, err
	}

	entry, err := lb.GetUserEntry(ctx, userID)
	if err != nil {
		return nil, err
	}
	if entry == nil {
		return &Entry{
			UserID:         userID,
			Score:          newScore,
			Metadata:       payload,
			TrimmedUserIDs: trimmedUserIDs,
		}, nil
	}

	entry.TrimmedUserIDs = trimmedUserIDs
	return entry, nil
}

// SetMaxEntries updates the leaderboard limit and trims immediately if needed.
func (lb *RedisLeaderboard) SetMaxEntries(ctx context.Context, maxEntries int) ([]string, error) {
	normalized, err := normalizePositiveInt(maxEntries, "maxEntries")
	if err != nil {
		return nil, err
	}

	lb.maxEntries = normalized
	return lb.trimToMaxEntries(ctx)
}

// GetTop returns the highest-ranked entries.
func (lb *RedisLeaderboard) GetTop(ctx context.Context, count int) ([]Entry, error) {
	normalized, err := normalizePositiveInt(count, "count")
	if err != nil {
		return nil, err
	}

	entries, err := lb.client.ZRevRangeWithScores(ctx, lb.key, 0, int64(normalized-1)).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to fetch top leaderboard entries: %w", err)
	}

	return lb.hydrateEntries(ctx, entries, 1)
}

// GetAroundRank returns entries centered around a 1-based rank.
func (lb *RedisLeaderboard) GetAroundRank(ctx context.Context, rank int, count int) ([]Entry, error) {
	normalizedRank, err := normalizePositiveInt(rank, "rank")
	if err != nil {
		return nil, err
	}
	normalizedCount, err := normalizePositiveInt(count, "count")
	if err != nil {
		return nil, err
	}

	totalEntries, err := lb.GetSize(ctx)
	if err != nil {
		return nil, err
	}
	if totalEntries == 0 {
		return []Entry{}, nil
	}
	if totalEntries <= int64(normalizedCount) {
		return lb.ListAll(ctx)
	}

	halfWindow := normalizedCount / 2
	start := max(0, normalizedRank-1-halfWindow)
	maxStart := int(totalEntries) - normalizedCount
	if start > maxStart {
		start = maxStart
	}
	end := start + normalizedCount - 1

	entries, err := lb.client.ZRevRangeWithScores(ctx, lb.key, int64(start), int64(end)).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to fetch leaderboard window: %w", err)
	}

	return lb.hydrateEntries(ctx, entries, start+1)
}

// GetRank returns a user's 1-based rank from the top.
func (lb *RedisLeaderboard) GetRank(ctx context.Context, userID string) (*int64, error) {
	rank, err := lb.client.ZRevRank(ctx, lb.key, userID).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to fetch leaderboard rank: %w", err)
	}

	normalizedRank := rank + 1
	return &normalizedRank, nil
}

// GetUserMetadata returns stored metadata for a user ID.
func (lb *RedisLeaderboard) GetUserMetadata(ctx context.Context, userID string) (map[string]string, error) {
	metadata, err := lb.client.HGetAll(ctx, lb.metadataKey(userID)).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to fetch user metadata: %w", err)
	}
	return metadata, nil
}

// GetUserEntry returns a single leaderboard entry with score, rank, and metadata.
func (lb *RedisLeaderboard) GetUserEntry(ctx context.Context, userID string) (*Entry, error) {
	pipe := lb.client.Pipeline()
	scoreCmd := pipe.ZScore(ctx, lb.key, userID)
	rankCmd := pipe.ZRevRank(ctx, lb.key, userID)
	metadataCmd := pipe.HGetAll(ctx, lb.metadataKey(userID))
	if _, err := pipe.Exec(ctx); err != nil && err != redis.Nil {
		return nil, fmt.Errorf("failed to fetch user entry: %w", err)
	}

	score, err := scoreCmd.Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to read user score: %w", err)
	}

	rank, err := rankCmd.Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to read user rank: %w", err)
	}

	metadata, err := metadataCmd.Result()
	if err != nil {
		return nil, fmt.Errorf("failed to read user metadata: %w", err)
	}

	return &Entry{
		Rank:     int(rank) + 1,
		UserID:   userID,
		Score:    score,
		Metadata: metadata,
	}, nil
}

// ListAll returns the full leaderboard from highest to lowest score.
func (lb *RedisLeaderboard) ListAll(ctx context.Context) ([]Entry, error) {
	entries, err := lb.client.ZRevRangeWithScores(ctx, lb.key, 0, -1).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to list leaderboard: %w", err)
	}

	return lb.hydrateEntries(ctx, entries, 1)
}

// GetSize returns the number of entries currently on the leaderboard.
func (lb *RedisLeaderboard) GetSize(ctx context.Context) (int64, error) {
	size, err := lb.client.ZCard(ctx, lb.key).Result()
	if err != nil {
		return 0, fmt.Errorf("failed to fetch leaderboard size: %w", err)
	}
	return size, nil
}

// DeleteUser removes a user from the leaderboard and deletes their metadata.
func (lb *RedisLeaderboard) DeleteUser(ctx context.Context, userID string) (bool, error) {
	pipe := lb.client.TxPipeline()
	zremCmd := pipe.ZRem(ctx, lb.key, userID)
	pipe.Del(ctx, lb.metadataKey(userID))
	if _, err := pipe.Exec(ctx); err != nil {
		return false, fmt.Errorf("failed to delete leaderboard user: %w", err)
	}

	removedCount, err := zremCmd.Result()
	if err != nil {
		return false, fmt.Errorf("failed to read delete result: %w", err)
	}

	return removedCount == 1, nil
}

// Clear removes all leaderboard scores and metadata for this leaderboard.
func (lb *RedisLeaderboard) Clear(ctx context.Context) error {
	userIDs, err := lb.client.ZRange(ctx, lb.key, 0, -1).Result()
	if err != nil {
		return fmt.Errorf("failed to fetch leaderboard members for clear: %w", err)
	}

	keys := make([]string, 0, len(userIDs)+1)
	keys = append(keys, lb.key)
	for _, userID := range userIDs {
		keys = append(keys, lb.metadataKey(userID))
	}

	return lb.client.Del(ctx, keys...).Err()
}

File: content/develop/use-cases/leaderboard/nodejs/leaderboard.js
/**
 * Redis-backed leaderboard implementation using a sorted set and user metadata hashes.
 */

class RedisLeaderboard {
  /**
   * @param {Object} options
   * @param {import("redis").RedisClientType} options.redisClient
   * @param {string} [options.key="leaderboard:demo"]
   * @param {number} [options.maxEntries=100]
   */
  constructor({ redisClient, key = "leaderboard:demo", maxEntries = 100 } = {}) {
    if (!redisClient) {
      throw new Error("redisClient is required");
    }

    this.redis = redisClient;
    this.key = key;
    this.maxEntries = this._normalizePositiveInt(maxEntries, "maxEntries");
  }

  _normalizePositiveInt(value, fieldName) {
    const normalized = Number.parseInt(value, 10);
    if (!Number.isFinite(normalized) || normalized < 1) {
      throw new Error(`${fieldName} must be at least 1`);
    }
    return normalized;
  }

  _metadataKey(userId) {
    return `${this.key}:user:${userId}`;
  }

  _coerceMetadata(metadata = null) {
    if (!metadata) {
      return {};
    }

    return Object.fromEntries(
      Object.entries(metadata).map(([field, value]) => [field, String(value)]),
    );
  }

  async _deleteMetadataForUsers(userIds) {
    if (!userIds.length) {
      return;
    }

    const keys = userIds.map((userId) => this._metadataKey(userId));
    await this.redis.sendCommand(["DEL", ...keys]);
  }

  async _zRangeWithScoresRev(start, end) {
    const response = await this.redis.sendCommand([
      "ZRANGE",
      this.key,
      String(start),
      String(end),
      "REV",
      "WITHSCORES",
    ]);

    const entries = [];
    for (let index = 0; index < response.length; index += 2) {
      entries.push({
        value: response[index],
        score: Number(response[index + 1]),
      });
    }

    return entries;
  }

  async _zRevRank(userId) {
    const response = await this.redis.sendCommand(["ZREVRANK", this.key, userId]);
    if (response === null) {
      return null;
    }

    return Number(response);
  }

  async _zScore(userId) {
    const response = await this.redis.sendCommand(["ZSCORE", this.key, userId]);
    if (response === null) {
      return null;
    }

    return Number(response);
  }

  async _hydrateEntries(entries, startRank) {
    if (!entries.length) {
      return [];
    }

    const multi = this.redis.multi();
    for (const entry of entries) {
      multi.hGetAll(this._metadataKey(entry.value));
    }
    const metadataResults = await multi.exec();

    return entries.map((entry, index) => ({
      rank: startRank + index,
      userId: entry.value,
      score: Number(entry.score),
      metadata: metadataResults[index] || {},
    }));
  }

  async _trimToMaxEntries() {
    const overflow = (await this.redis.zCard(this.key)) - this.maxEntries;
    if (overflow <= 0) {
      return [];
    }

    const trimmedUserIds = await this.redis.zRange(this.key, 0, overflow - 1);
    if (!trimmedUserIds.length) {
      return [];
    }

    await this.redis.zRemRangeByRank(this.key, 0, overflow - 1);
    await this._deleteMetadataForUsers(trimmedUserIds);
    return trimmedUserIds;
  }

  async upsertUser(userId, score, metadata = null) {
    const metadataKey = this._metadataKey(userId);
    const payload = this._coerceMetadata(metadata);

    const multi = this.redis.multi();
    multi.zAdd(this.key, [{ score: Number(score), value: userId }]);
    if (Object.keys(payload).length > 0) {
      multi.hSet(metadataKey, payload);
    }
    await multi.exec();

    const trimmedUserIds = await this._trimToMaxEntries();
    const entry = await this.getUserEntry(userId);
    if (!entry) {
      return {
        userId,
        score: Number(score),
        metadata: payload,
        trimmedUserIds,
      };
    }

    return {
      ...entry,
      trimmedUserIds,
    };
  }

  async incrementScore(userId, amount, metadata = null) {
    const metadataKey = this._metadataKey(userId);
    const payload = this._coerceMetadata(metadata);

    const multi = this.redis.multi();
    multi.zIncrBy(this.key, Number(amount), userId);
    if (Object.keys(payload).length > 0) {
      multi.hSet(metadataKey, payload);
    }
    const result = await multi.exec();

    const newScore = Number(result[0]);
    const trimmedUserIds = await this._trimToMaxEntries();
    const entry = await this.getUserEntry(userId);
    if (!entry) {
      return {
        userId,
        score: newScore,
        metadata: payload,
        trimmedUserIds,
      };
    }

    return {
      ...entry,
      trimmedUserIds,
    };
  }

  async setMaxEntries(maxEntries) {
    this.maxEntries = this._normalizePositiveInt(maxEntries, "maxEntries");
    return this._trimToMaxEntries();
  }

  async getTop(count) {
    const normalizedCount = this._normalizePositiveInt(count, "count");
    const entries = await this._zRangeWithScoresRev(0, normalizedCount - 1);
    return this._hydrateEntries(entries, 1);
  }

  async getAroundRank(rank, count) {
    const normalizedRank = this._normalizePositiveInt(rank, "rank");
    const normalizedCount = this._normalizePositiveInt(count, "count");
    const totalEntries = await this.getSize();

    if (totalEntries === 0) {
      return [];
    }

    if (totalEntries <= normalizedCount) {
      return this.listAll();
    }

    const halfWindow = Math.floor(normalizedCount / 2);
    let start = Math.max(0, normalizedRank - 1 - halfWindow);
    start = Math.min(start, totalEntries - normalizedCount);
    const end = start + normalizedCount - 1;

    const entries = await this._zRangeWithScoresRev(start, end);
    return this._hydrateEntries(entries, start + 1);
  }

  async getRank(userId) {
    const rank = await this._zRevRank(userId);
    if (rank === null) {
      return null;
    }
    return rank + 1;
  }

  async getUserMetadata(userId) {
    return this.redis.hGetAll(this._metadataKey(userId));
  }

  async getUserEntry(userId) {
    const [score, rank, metadata] = await Promise.all([
      this._zScore(userId),
      this._zRevRank(userId),
      this.redis.hGetAll(this._metadataKey(userId)),
    ]);

    if (score === null || rank === null) {
      return null;
    }

    return {
      rank: rank + 1,
      userId,
      score,
      metadata: metadata || {},
    };
  }

  async listAll() {
    const entries = await this._zRangeWithScoresRev(0, -1);
    return this._hydrateEntries(entries, 1);
  }

  async getSize() {
    return Number(await this.redis.zCard(this.key));
  }

  async deleteUser(userId) {
    const multi = this.redis.multi();
    multi.zRem(this.key, userId);
    multi.del(this._metadataKey(userId));
    const [removedCount] = await multi.exec();
    return Boolean(removedCount);
  }

  async clear() {
    const userIds = await this.redis.zRange(this.key, 0, -1);
    const keys = [this.key, ...userIds.map((userId) => this._metadataKey(userId))];
    await this.redis.sendCommand(["DEL", ...keys]);
  }
}

module.exports = { RedisLeaderboard };
